Mathlib/Combinatorics/SimpleGraph/Hamiltonian.lean

```diff
@@ -0,0 +1,138 @@
+def IsHamiltonian (p : G.Walk a b) : Prop := ∀ a, p.support.count a = 1
+
+lemma IsHamiltonian.map {H : SimpleGraph β} (f : G →g H) (hf : Bijective f) (hp : p.IsHamiltonian) :
+    (p.map f).IsHamiltonian := by
+  simp [IsHamiltonian, hf.surjective.forall, hf.injective, hp _]
+
+/-- A hamiltonian path visits every vertex. -/
+@[simp] lemma IsHamiltonian.mem_support (hp : p.IsHamiltonian) (c : α) : c ∈ p.support := by
+  simp only [← List.count_pos_iff_mem, hp _, Nat.zero_lt_one]
+
+/-- The support of a hamiltonian walk is the entire vertex set. -/
+lemma IsHamiltonian.support_toFinset (hp : p.IsHamiltonian) : p.support.toFinset = Finset.univ := by
+  simp [eq_univ_iff_forall, hp]
+
+/-- The length of a hamiltonian path is one less than the number of vertices of the graph. -/
+lemma IsHamiltonian.length_eq (hp : p.IsHamiltonian) : p.length = Fintype.card α - 1 :=
+  eq_tsub_of_add_eq $ by
+    rw [← length_support, ← List.sum_toFinset_count_eq_length, Finset.sum_congr rfl fun _ _ ↦ hp _,
+      ← card_eq_sum_ones, hp.support_toFinset, card_univ]
+
+/-- Hamiltonian paths are paths. -/
+lemma IsHamiltonian.isPath (hp : p.IsHamiltonian) : p.IsPath :=
+  IsPath.mk' <| List.nodup_iff_count_le_one.2 <| (le_of_eq <| hp ·)
+
+/-- A path whose support contains every vertex is hamiltonian. -/
+lemma IsPath.isHamiltonian_of_mem (hp : p.IsPath) (hp' : ∀ w, w ∈ p.support) :
+    p.IsHamiltonian := fun _ ↦
+  le_antisymm (List.nodup_iff_count_le_one.1 hp.support_nodup _) (List.count_pos_iff_mem.2 (hp' _))
+
+lemma IsPath.isHamiltonian_iff (hp : p.IsPath) : p.IsHamiltonian ↔ ∀ w, w ∈ p.support :=
+  ⟨(·.mem_support), hp.isHamiltonian_of_mem⟩
+
+/-- A hamiltonian cycle is a cycle that visits every vertex once. -/
+structure IsHamiltonianCycle (p : G.Walk a a) extends p.IsCycle : Prop :=
+  isHamiltonian_tail : (p.tail toIsCycle.not_nil).IsHamiltonian
+
+variable {p : G.Walk a a}
+
+lemma IsHamiltonianCycle.isCycle (hp : p.IsHamiltonianCycle) : p.IsCycle :=
+  hp.toIsCycle
+
+lemma IsHamiltonianCycle.map {H : SimpleGraph β} (f : G →g H) (hf : Bijective f)
+    (hp : p.IsHamiltonianCycle) : (p.map f).IsHamiltonianCycle where
+  toIsCycle := by
+    rw [map_isCycle_iff_of_injective]
+    exact isCycle hp
+    exact Bijective.injective hf
+  isHamiltonian_tail := by
+    simp [IsHamiltonian, support_tail, hf.surjective.forall, List.count_tail, hf.injective]
+    intro x
+    rcases p with (_ | ⟨y, p⟩)
+    · cases hp.ne_nil rfl
+    simp only [support_cons, List.count_cons, List.map_cons, List.head_cons, hf.injective.eq_iff,
+      add_tsub_cancel_right]
+    exact hp.isHamiltonian_tail _
+
+lemma IsHamiltonianCycle_def :
+    p.IsHamiltonianCycle ↔ ∃ h : p.IsCycle, (p.tail h.not_nil).IsHamiltonian :=
+  ⟨fun ⟨h, h'⟩ ↦ ⟨h, h'⟩, fun ⟨h, h'⟩ ↦ ⟨h, h'⟩⟩
+
+lemma IsHamiltonianCycle_iff :
+    p.IsHamiltonianCycle ↔ p.IsCycle ∧ ∀ a, (support p).tail.count a = 1 := by
+  simp only [IsHamiltonianCycle_def, IsHamiltonian, support_tail, exists_prop]
```

Comment:
Not super clear which one should be the `_def`. Maybe rename to `isHamiltonianCycle_isCycle_and_isHamiltonian_tail` and `isHamiltonianCycle_iff_isCycle_and_support_count_tail_eq_one`?

Also note that `IsHamiltonianCycle_something` is wrong. It must be in `lowerCamelCase`. Compare this to `IsHamiltonianCycle.something`, where it is in `UpperCamelCase` (and `something` is in `lowerCamelCase`).

Reply:
I like these names a lot better as well. Done.